### DISCUSSION TITLE
Running e2e tests on Vercel previews

### DISCUSSION THREAD

**Discussion by baptisteArno** [nextauthjs/next-auth#3765] [Help] [answered]
I'm running e2e tests (Playwright) on Vercel previews when a new version is deployed.

Now comes the part where I need to signing and I'm king of blocked because my enabled OAuth providers (Google, GitHub, Facebook) don't allow dynamic return URLs and it seems there's no way to bypass the email provider.

I can't use Credentials provider either because I'm using `database` strategy.

What should I do here?

**Comment by baptisteArno** [comment]
Hey @balazsorban44, sorry to ping you. I'm just completely blocked on this issue.

Do you have a clue?

**Comment by balazsorban44** [comment, marked as answer]
Thanks for the ping. Unfortunately, there is not much NextAuth.js can do about it, since this is a limitation of OAuth providers (be it unintentional or not).

We recently created some guides on how to authenticate users in preview environments, I hope that helps: https://next-auth.js.org/deployment#securing-a-preview-deployment

Basically, you could use an environment variable to toggle between allowed providers, or even switch off the adapter when you are in a test/preview environment. Making sure that the actual login flow with OAuth will work is our library's responsibility, not your app's.

Otherwise, you could try to use something like [MSW](https://mswjs.io) and mock the requests that NextAuth.js makes. Especially the one to `/api/auth/session`. https://next-auth.js.org/getting-started/rest-api

**Reply by baptisteArno** [reply]
Thank you so much for the answer, everything makes sense now 👍

**Reply by balazsorban44** [reply]
To add, this is also one of the reasons we cannot test our provider integrations automatically (#2524). CI/CD environments are considered bots (rightfully so) and therefore providers will often show a captcha/challenge that makes their login process impossible in tests. So mocking/skipping really is the only option as far as I am aware.

**Reply by balazsorban44** [reply]
If you think this is useful, would you care about opening a PR to add this to the docs so we can reference it in the future? :pray:

**Reply by baptisteArno** [reply]
I'll give MSW a try and when everything works on my end, I'll create a PR 👍
